doc/tutorials/debug.rst

```diff
@@ -0,0 +1,322 @@
+
+5. Check NPK is enabled in AaaG::
+
+      $ ls /sys/bus/pci/drivers/intel_th_pci/
+      0000:00:00.2 bind new_id remove_id uevent unbind
+
+6. Enable userspace log of UOS (Android) into NPK::
+
+      $ setprop vendor.npk.cfg.update pstore,android
+
+7. Enable kernel log of UOS (Android) into NPK::
+
+      $ setprop vendor.npk.cfg.update pstore,kernel
+
+8. Enable NPK log output to memory using the following command in SOS::
+
+      $ echo 0 > /sys/bus/intel_th/devices/0-gth/masters/0
+      $ echo 0 > /sys/bus/intel_th/devices/0-gth/masters/74
+      $ echo 0 > /sys/bus/intel_th/devices/0-gth/masters/256+
+      $ echo 1024,1024,1024 > /sys/bus/intel_th/devices/0-msc0/nr_pages
+
+Set and grab log
+================
+
+We have 1-6 log levels for console log and mem log. The following
+functions print different levels of console log and mem log::
+
+      pr_dbg("debug...level %d", LOG_DEBUG);       //level 6
+      pr_info("info...level %d", LOG_INFO);        //level 5
+      pr_warn("warn...level %d", LOG_WARNING);     //level 4
+      pr_err("err...level %d", LOG_ERROR);         //level 3
+      pr_acrnlog("acrnlog...level %d", LOG_ACRN);  //level 2
+      pr_fatal("fatal...level %d", LOG_FATAL);     //level 1
+
+If the built-in logging doesn't provide enough information, you can add
+additional logging in functions you want to debug, using the functions
+noted above.  For example, add the following code into function
+``shell_cmd_help`` in the source file
+``acrn-hypervisor/hypervisor/debug/shell.c``:
+
+.. figure:: images/debug_image22.png
+   :align: center
+
+   shell_cmd_help added information
+
+We set console log level to 5, and mem log level to 2 through the
+command::
+
+   loglevel 5 2
+
+Then we input "help" into the ACRN console, and find the log as follows.
+
+.. figure:: images/debug_image23.png
+   :align: center
+
+   console log information
+
+Then we use the command, on the ACRN console::
+
+   sos_console
+
+to switch to the SOS console the use the commands::
+
+   systemctl enable acrnlog
+   reboot
+
+to enable the mem log and reboot the SOS.
+
+Next we set the console and mem log levels with::
+
+   loglevel 5 2
+
+and run the command::
+
+   help
+
+Finally we get into the SOS and use the command::
+
+   cat /tmp/acrnlog/acrnlog_cur.0
+
+on the console, and we'll see the following log:
+
+.. figure:: images/debug_image24.png
+   :align: center
+
+   mem log information
+
+When we have enabled NPK in the hypervisor, SOS, and UOS, we can use the
+following commands to get the NPK log:
+
+1. Start NPK output in SOS::
+
+      $ echo 1 > /sys/bus/intel_th/devices/0-msc0/active
+
+2. Wait a moment, stop NPK output, and get the log from memory::
+
+      $ echo 0 > /sys/bus/intel_th/devices/0-msc0/active
+      $ cat /dev/intel_th0/msc0 > npk_trace
+
+3. Use `PVT
+   <https://ubit-artifactory.intel.com/artifactory/list/dsdjf-pvt-repo/1909/Red/InstallPVT.All.1.0.9006.600.exe>`_
```

Comment:
That's not widely accessible, is it?

Reply:
yes, so it is the same problem as the above question.